### PR TITLE
BFD-2876:  Upload Test Results to SonarQube

### PR DESCRIPTION
**Pull request by aadedejifearless**
**JIRA Ticket:**
[BFD-2876](https://jira.cms.gov/browse/BFD-2876)

**User Story or Bug Summary:**

---
As a BFD contributor I want the test coverage in SonarQube to be accurate so that I can write tests to improve coverage.

### What Does This PR Do?

- Updates the `build.groovy` file
- Updates the `JenkinsFile` in `ops/jenkins/bfd-jenkins-sonarqube/Jenkinsfile`
- Updates `apps/pom.xml` to update sonar jacoco path. 

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [X] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [X] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [X] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [X] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [X] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [X] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [X] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [X] The data dictionary has been updated with any field mapping changes, if any were made.
* [X] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [X] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [X] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [X] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/pom.xml [hunk 1]
```diff
@@ -863,7 +863,7 @@
                 <plugin>
                     <groupId>org.apache.maven.plugins</groupId>
                     <artifactId>maven-site-plugin</artifactId>
-                    <version>3.4</version>
```
Comment (billie-rose) [inline apps/pom.xml:866]: Does this plugin need to be updated for this to work?

File: apps/pom.xml [hunk 2]
```diff
@@ -933,7 +933,7 @@
             <plugin>
                 <groupId>org.codehaus.mojo</groupId>
                 <artifactId>versions-maven-plugin</artifactId>
-                <version>2.11.0</version>
+                <version>2.16.1</version>
```
Comment (billie-rose) [inline apps/pom.xml:936]: Does this plugin need to be updated for this to work?

File: apps/pom.xml (hunk 1 above)
Comment (aadedejifearless) [reply apps/pom.xml]: Reverted this change. This doesnt need to change for anything to work. This was done to supress the warnings from the plugin.validation.

File: apps/pom.xml (hunk 2 above)
Comment (aadedejifearless) [reply apps/pom.xml:936]: This upgrade removed the validation warning for this particular. The upgrade was done to remove these warnings as these versions might not be compatible in the future with Maven version 4.0+. I reverted the previous comment as that did not remove the validation warnings.

```
[WARNING] 
[WARNING] Plugin validation issues were detected in 3 plugin(s)
[WARNING] 
[WARNING]  * org.apache.maven.plugins:maven-site-plugin:3.4
[WARNING]  * org.codehaus.mojo:versions-maven-plugin:2.11.0
[WARNING]  * org.xolstice.maven.plugins:protobuf-maven-plugin:0.6.1
[WARNING] 
```

Comment (mjburling) [reply apps/pom.xml]: I think this is a really good intention, but as discussed, let's keep this PR focused on the changes that help us realize code coverage report generation/upload. The various upgrades can be handled in [BFD-2770](https://jira.cms.gov/browse/BFD-2770). Might be worthwhile to comment on that story to call this specific upgrade out 👍

**Review by mjburling — APPROVED** [review]
🥇

File: apps/pom.xml [hunk 3]
```diff
@@ -75,8 +75,8 @@
         <maven.jacoco.skip>true</maven.jacoco.skip>
         <sonar.java.coveragePlugin>jacoco</sonar.java.coveragePlugin>
         <sonar.dynamicAnalysis>reuseReports</sonar.dynamicAnalysis>
-        <sonar.jacoco.reportPath>${project.build.directory}/coverage-reports/jacoco-ut.exec</sonar.jacoco.reportPath>
-        <sonar.jacoco.itReportPath>${project.build.directory}/coverage-reports/jacoco-it.exec</sonar.jacoco.itReportPath>
+        <sonar.coverage.jacoco.xmlReportPaths>${project.build.directory}/target/coverage-reports/jacoco-ut/jacoco.xml</sonar.coverage.jacoco.xmlReportPaths>
+        <sonar.coverage.jacoco.itReportPath>${project.build.directory}/target/coverage-reports/jacoco-it/jacoco.xml</sonar.coverage.jacoco.itReportPath>
```
Comment (billie-rose) [inline apps/pom.xml:79]: Lines 78 & 79 - do we need `/target/`? The  plugin config has its `outputDirectories` set to
`<outputDirectory>${project.build.directory}/coverage-reports/jacoco-it</outputDirectory>`
`<outputDirectory>${project.build.directory}/coverage-reports/jacoco-ut</outputDirectory>`

Comment (aadedejifearless) [reply apps/pom.xml:79]: Let me run it without and see ...

File: ops/jenkins/bfd-jenkins-sonarqube/Jenkinsfile [hunk 4]
```diff
@@ -1,8 +1,12 @@
 #!/usr/bin/env groovy
 
 // These variables are accessible throughout this file (except inside methods and classes).
-def verboseMaven = params.verbose_mvn_logging
-def appBuildResults
+def verboseMaven
```
Comment (billie-rose) [inline ops/jenkins/bfd-jenkins-sonarqube/Jenkinsfile:4]: I *think* `null` is a falsey value in groovy already, so this shouldn't be necessary.

Comment (mjburling) [reply ops/jenkins/bfd-jenkins-sonarqube/Jenkinsfile:4]: > I _think_ `null` is a falsey value in groovy already, so this shouldn't be necessary.

I'm no expert, I just know that assigning variables to e.g. `params.foo` **always** fails on the first run in Jenkins. This _trashy_ little solution helps us avoid the first-run failure that impacted this and other jobs. Have you seen this similar behavior in e.g. [bfd-build-apps](https://jenkins-east.cloud.cms.gov/bfd/view/Build/job/bfd-build-apps/)?

Comment (billie-rose) [reply ops/jenkins/bfd-jenkins-sonarqube/Jenkinsfile:4]: Ope, yes. Disregard then :)

**Review by billie-rose — APPROVED** [review]

File: apps/pom.xml [hunk 5]
```diff
@@ -75,8 +75,8 @@
         <maven.jacoco.skip>true</maven.jacoco.skip>
         <sonar.java.coveragePlugin>jacoco</sonar.java.coveragePlugin>
         <sonar.dynamicAnalysis>reuseReports</sonar.dynamicAnalysis>
-        <sonar.jacoco.reportPath>${project.build.directory}/coverage-reports/jacoco-ut.exec</sonar.jacoco.reportPath>
-        <sonar.jacoco.itReportPath>${project.build.directory}/coverage-reports/jacoco-it.exec</sonar.jacoco.itReportPath>
+        <sonar.coverage.jacoco.xmlReportPaths>${project.build.directory}/coverage-reports/jacoco-ut/jacoco.xml</sonar.coverage.jacoco.xmlReportPaths>
+        <sonar.coverage.jacoco.itReportPath>${project.build.directory}/coverage-reports/jacoco-it/jacoco.xml</sonar.coverage.jacoco.itReportPath>
```
Comment (brianburton) [inline apps/pom.xml:79]: Is this really correct?  Our unit and integration tests write jacoco data to two different exec files and they were also writing their XML output to two different XML files.

Comment (aadedejifearless) [reply apps/pom.xml:79]: Sonar no longers supports `jacoco.exec` because its in binary format instead of being in XML.

**Pull request merged**